### PR TITLE
Improve header handling

### PR DESCRIPTION
**Pull request by okuta**
This PR proves to use `#include` correctly.
Please merge this PR before #214.

### CODE REVIEW

File: cupy/cuda/compiler.py [hunk 1]
```diff
@@ -81,17 +80,17 @@ def compile_using_nvrtc(source, options=(), arch=None):
...
     except CompileException as e:
         dump = _get_bool_env_variable(
             'CUPY_DUMP_CUDA_SOURCE_ON_ERROR', False)
         if dump:
             e.dump(sys.stderr)
         raise
 
-    if isinstance(pp_src, six.binary_type):
-        pp_src = pp_src.decode('utf-8')
-    return re.sub('(?m)^#.*$', '', pp_src)
```
Comment (unnonouno) [inline cupy/cuda/compiler.py]: Why is this removed?

File: cupy/core/carray.pxi [hunk 2]
```diff
@@ -60,19 +60,41 @@ cdef class Indexer:
         return CIndexer(self.size, self.shape)
 
 
+cdef list _cupy_header_list = [
+    'cupy/carray.cuh',
+]
+cdef str _cupy_header = """
```
Comment (unnonouno) [inline cupy/core/carray.pxi]: It can be generated from `_cupy_header_list`.

File: cupy/core/carray.pxi [hunk 3]
```diff
@@ -60,19 +60,41 @@ cdef class Indexer:
...
+#include <cupy/carray.cuh>
+"""
+
+cdef str _header_path_cache = None
```
Comment (okuta) [inline cupy/core/carray.pxi]: Unfortunately, we cannot use `__file__` in module level.

File: cupy/cuda/compiler.py (hunk 1 above)
Comment (okuta) [inline cupy/cuda/compiler.py]: This replacing is not used now.

File: cupy/core/carray.pxi [hunk 4]
```diff
@@ -60,19 +60,40 @@ cdef class Indexer:
         return CIndexer(self.size, self.shape)
 
 
+cdef list _cupy_header_list = [
+    'cupy/carray.cuh',
+]
+cdef str _cupy_header = '\n'.join(
+    ['#include <%s>' % i for i in _cupy_header_list])
```
Comment (unnonouno) [inline cupy/core/carray.pxi]: It is a bit different from the original. 
```
cdef str _cupy_header = ''.join(
    ['#include <%s>\n' % i for i in _cupy_header_list])
```

Comment (okuta) [inline cupy/core/carray.pxi]: Thanks, I fixed.

**Comment by okuta** [comment]
I fixed.

File: cupy/cuda/compiler.py [hunk 5]
```diff
@@ -103,7 +102,8 @@ def get_cache_dir():
 _empty_file_preprocess_cache = {}
 
 
-def compile_with_cache(source, options=(), arch=None, cache_dir=None):
+def compile_with_cache(source, options=(), arch=None, cache_dir=None,
+                       extra_source=None):
```
Comment (unnonouno) [inline cupy/cuda/compiler.py]: Write a comment why `extra_source` is required. It is because we cannot use nvrtc to get result of preprocess though want to check if header files are changed.

Comment (okuta) [inline cupy/cuda/compiler.py]: I added

**Comment by unnonouno** [comment]
LGTM
